homeassistant/components/climate/generic_thermostat.py

```diff
@@ -129,17 +155,70 @@ def current_operation(self):
             return STATE_HEAT if heating else STATE_IDLE
 
     @property
+    def operation_list(self):
+        """List of available operation modes."""
+        return self._operation_list
+
+    @property
+    def current_hold_mode(self):
+        """Return the current hold mode."""
+        return self._current_hold_mode
+
+    @property
+    def hold_list(self):
+        """List of available hold modes."""
+        # Support all hold modes, as they all are software defined.
+        return HOLD_MODE_LIST
+
+    @asyncio.coroutine
+    def async_set_hold_mode(self, hold_mode):
+        """Set hold mode (currently understood modes: away and home)."""
+        if hold_mode == self.current_hold_mode:
+            # no change, so no action required
+            return
+
+        if hold_mode not in HOLD_MODE_LIST:
+            _LOGGER.warning('Hold mode `%s` is not supported, ignoring it.',
+                            hold_mode)
+            return
+
+        if not self._hold_temps.get(hold_mode):
+            _LOGGER.warning('Hold mode `%s` has no temperature set, ignoring',
+                            hold_mode)
+            return
+
+        # store the (known) mode
+        self._current_hold_mode = hold_mode
+        _LOGGER.info('The hold mode has been set to `%s`.', hold_mode)
+        # and force a refresh
+        self._async_control_heating()
+        yield from self.async_update_ha_state()
+
+    @property
     def target_temperature(self):
-        """Return the temperature we try to reach."""
+        """Return the temperature we try to reach (in default "home" mode)."""
         return self._target_temp
 
     @asyncio.coroutine
     def async_set_temperature(self, **kwargs):
         """Set new target temperature."""
         temperature = kwargs.get(ATTR_TEMPERATURE)
-        if temperature is None:
+        hold_mode = kwargs.get(ATTR_HOLD_MODE, HOLD_MODE_HOME)
+
+        if hold_mode not in HOLD_MODE_LIST:
+            _LOGGER.warning('Hold mode `%s` is not supported, ignoring it.',
+                            hold_mode)
             return
-        self._target_temp = temperature
+
+        if hold_mode == HOLD_MODE_HOME:
+            # cannot set None to the default temperature
+            if temperature is None:
+                return
+            self._target_temp = temperature
+
+        # Store the temperature in the hold_temps dictionary
+        self._hold_temps[hold_mode] = temperature
```

Comment:
Log error and ignore is good, until we fix this in the service schema properly. You should still also add the hold mode attribute as an optional attribute in the service schema.